### PR TITLE
Optimize _lwt_AddFaceSplit short circuit contains with bbox

### PR DESCRIPTION
**Pull request by bjornharrtell**
In a case of loading a pretty large topology via first adding nodes then all edges with ST_AddEdgeNewFaces in ported code it speeds up the process by more than 10x (!). That said, performance implications might be entirely different in in this implementation.

The rationale is that the looped edges can be quite large in case of loading lots of edges into universe forming faces as we go and doing proper contains check for all can be a hot path.

### CODE REVIEW

**Comment by bjornharrtell** [comment]
"psql:getfacecontainingpoint.sql:21: server closed the connection unexpectedly" :scream:

File: liblwgeom/topo/lwgeom_topo.c [hunk 1]
```diff
@@ -2141,9 +2141,8 @@ _lwt_AddFaceSplit( LWT_TOPOLOGY* topo,
         return -2;
       }
 
-      /* IDEA: check that bounding box shortcut is taken, or use
-       *       shellbox to do it here */
-      contains = ptarray_contains_point(pa, &ep);
+      contains = gbox_contains_point2d(newface.mbr, &ep) == LW_TRUE ? LW_INSIDE : LW_OUTSIDE;
```
Comment (strk) [inline liblwgeom/topo/lwgeom_topo.c:2144]: should you use "shellbox" here instead of "newface.mbr" ? Just to follow the IDEA comment, didn't really check the code (yes I know it's hard to follow)

Comment (bjornharrtell) [reply liblwgeom/topo/lwgeom_topo.c]: Ah yes :) seems to work better now. In a case of loading a pretty large topology via first adding nodes then all edges with ST_AddEdgeNewFaces in ported code it speeds up the process by more than 10x (!).

**Comment by strk** [comment]
Could you share any numbers ? Maybe topology/test/perf can help with that

**Comment by strk** [comment]
@larsop do you have a large testcase to test this change against ? It's supposed to speed topology building up when incoming lines close new rings whose bounding box overlaps with many non-trivial edges

**Comment by bjornharrtell** [comment]
I did try it on a medium sized set and did not observe expected speedups, unfortunately.

**Comment by larsop** [comment]
> @larsop do you have a large testcase to test this change against ? It's supposed to speed topology building up when incoming lines close new rings whose bounding box overlaps with many non-trivial edges

Yes I can try to compile and test https://github.com/bjornharrtell/postgis/tree/addfacesplit-bbox-opt, but first I have to fix this one  https://gitlab.com/nibioopensource/resolve-overlap-and-gap/-/issues/90 . Here the problem was that I silently got an invalid topology. I almost have a test ready I will try report tonight. Hopefully I get the time to test addfacesplit-bbox-opt this weekend, I need some servers to run on. I use a lot of time when adding lines crossing cell borders to merge tmp topologies into big master topology, should this also speed up this operation ?

**Comment by larsop** [comment]
I use topology.TopoGeo_addLinestring I assume that should be affected by this fix or do I have do any changes in may code ?

**Comment by strk** [comment]
> I use topology.TopoGeo_addLinestring I assume that should be affected by this fix or do I have do any changes in may code ?

I confirm this change will affect `TopoGeo_addLinestring` when the added line closes a new ring

**Comment by larsop** [comment]
Sorry, did not see any big performance gains here. I will run bigger tests through the weekend.

'topology tmp_dyrkbarjord_11_*' is `POSTGIS="3.6.0dev 3.5.0-24-g26eb9730c" [EXTENSION] PGSQL="160" GEOS="3.13.0-CAPI-1.19.0" PROJ="8.2.1 NETWORK_ENABLED=OFF URL_ENDPOINT=https://cdn.proj.org USER_WRITABLE_DIRECTORY=/tmp/proj DATABASE_PATH=/usr/share/proj/proj.db" (compiled against PROJ 8.2.1) LIBXML="2.9.13" LIBJSON="0.13.1" LIBPROTOBUF="1.3.3" WAGYU="0.5.0 (Internal)" TOPOLOGY
`

'topology tmp_dyrkbarjord_12_*' is `POSTGIS="3.6.0dev 6404944a2" [EXTENSION] PGSQL="160" GEOS="3.13.0-CAPI-1.19.0" PROJ="8.2.1 NETWORK_ENABLED=OFF URL_ENDPOINT=https://cdn.proj.org USER_WRITABLE_DIRECTORY=/tmp/proj DATABASE_PATH=/usr/share/proj/proj.db" (compiled against PROJ 8.13.0) LIBXML="2.9.13" LIBJSON="0.13.1" LIBPROTOBUF="1.3.3" WAGYU="0.5.0 (Internal)" TOPOLOGY
`

Total runtime with tmp_dyrkbarjord_11_* was `Time: 1987106.052 ms (33:07.106)`

Total runtime with tmp_dyrkbarjord_12_* was `Time: 1876919.213 ms (31:16.919)`

Here some clips from the logs

```
NOTICE:  Done with 220 number of add_border_lines for _box_id 93 to temp topollogy (number of failed lines <NULL>), number edges to remove 1144, phase 0 for topology tmp_dyrkbarjord_11_037_93  with tolerance 0 at 2024-10-25 16:05:06.120126+02 used_time: 116.26169
NOTICE:  Done with 220 number of add_border_lines for _box_id 93 to temp topollogy (number of failed lines <NULL>), number edges to remove 1144, phase 0 for topology tmp_dyrkbarjord_12_037_93  with tolerance 0 at 2024-10-25 17:31:18.547394+02 used_time: 117.64703
NOTICE:  Done with 219 number of add_border_lines for _box_id 94 to temp topollogy (number of failed lines <NULL>), number edges to remove 1365, phase 0 for topology tmp_dyrkbarjord_11_037_94  with tolerance 0 at 2024-10-25 16:05:15.884553+02 used_time: 123.058655
NOTICE:  Done with 219 number of add_border_lines for _box_id 94 to temp topollogy (number of failed lines <NULL>), number edges to remove 1365, phase 0 for topology tmp_dyrkbarjord_12_037_94  with tolerance 0 at 2024-10-25 17:31:05.786974+02 used_time: 104.74367
NOTICE:  Done with 451 number of add_border_lines for _box_id 95 to temp topollogy (number of failed lines <NULL>), number edges to remove 3180, phase 0 for topology tmp_dyrkbarjord_11_037_95  with tolerance 0 at 2024-10-25 16:08:01.518168+02 used_time: 290.08603
NOTICE:  Done with 451 number of add_border_lines for _box_id 95 to temp topollogy (number of failed lines <NULL>), number edges to remove 3180, phase 0 for topology tmp_dyrkbarjord_12_037_95  with tolerance 0 at 2024-10-25 17:33:44.887802+02 used_time: 267.69806
NOTICE:  Done with 324 number of add_border_lines for _box_id 96 to temp topollogy (number of failed lines <NULL>), number edges to remove 1167, phase 0 for topology tmp_dyrkbarjord_11_037_96  with tolerance 0 at 2024-10-25 16:05:23.568016+02 used_time: 135.48619
NOTICE:  Done with 324 number of add_border_lines for _box_id 96 to temp topollogy (number of failed lines <NULL>), number edges to remove 1167, phase 0 for topology tmp_dyrkbarjord_12_037_96  with tolerance 0 at 2024-10-25 17:31:26.453897+02 used_time: 127.35823

```

**Comment by bjornharrtell** [comment]
Well, at least it's a bit faster. :) There is probably another bottleneck that I do not have in my port (which is currently a RAM only implementation).

**Comment by larsop** [comment]
Yes that's good, thanks for checking up this and just let me know and when you need more testing.

Another thing here is that I add my lines in a very controlled way  based on what just seemed work to best over the years. 
When added data to each temp topology  I do `ORDER BY r.table_input_order,  ST_IsClosed(r.geom) desc, ST_Length(r.geom,true) ASC` . Maybe this explains way we do not see the big differences.

After the temp topologies are added to the master I need to connect the borders crossing the cell borders and I then use another sorting order  after r.table_input_order. I first try add lines that touches added edges on both sides (yes there edges not  touching on both sides), after this I sort on  `ST_X(ST_StartPoint(r.geom)), ST_Y(ST_StartPoint(r.geom))`

Here clip from the log when adding cell crossing edges from the same test as above.

```
NOTICE:  In add_cellborder_lines, added border 502 lines in this loop  (total all loops 260) (total to add 1137) used_time 15.114047 secs. (edges created 502) for tmp_dyrkbarjord_11_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 501 lines in this loop  (total all loops 486) (total to add 1137) used_time 24.532326 secs. (edges created 1003) for tmp_dyrkbarjord_11_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 502 lines in this loop  (total all loops 719) (total to add 1137) used_time 19.78274 secs. (edges created 1505) for tmp_dyrkbarjord_11_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 535 lines in this loop  (total all loops 900) (total to add 1137) used_time 20.573816 secs. (edges created 2040) for tmp_dyrkbarjord_11_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 502 lines in this loop  (total all loops 1100) (total to add 1137) used_time 11.238217 secs. (edges created 2542) for tmp_dyrkbarjord_11_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 511 lines in this loop  (total all loops 256) (total to add 1137) used_time 15.839447 secs. (edges created 511) for tmp_dyrkbarjord_12_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 502 lines in this loop  (total all loops 492) (total to add 1137) used_time 22.186863 secs. (edges created 1013) for tmp_dyrkbarjord_12_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 501 lines in this loop  (total all loops 704) (total to add 1137) used_time 22.047363 secs. (edges created 1514) for tmp_dyrkbarjord_12_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 501 lines in this loop  (total all loops 926) (total to add 1137) used_time 14.831797 secs. (edges created 2015) for tmp_dyrkbarjord_12_037 bbox area <NULL> centroid <NULL> .
NOTICE:  In add_cellborder_lines, added border 501 lines in this loop  (total all loops 1096) (total to add 1137) used_time 16.28515 secs. (edges created 2516) for tmp_dyrkbarjord_12_037 bbox area <NULL> centroid <NULL> .

```

**Comment by larsop** [comment]
Below I have done some tests with some bigger topologies running on another server where set the limit on 2000 polygons pr grid cell , in the first test I had 600 so this will make each temp topology bigger.

The gronn_2023_v2_004* is with out your patch and gronn_2023_v3_004* is with.

```
NOTICE:  Done with 2336 number of add_border_lines for _box_id 15 to temp topollogy (number of failed lines <NULL>), number edges to remove 2452, phase 0 for topology gronn_2023_v2_004_15  with tolerance 0 at 2024-10-25 10:33:03.277983+02 used_time: 131.22925
NOTICE:  Done with 2336 number of add_border_lines for _box_id 15 to temp topollogy (number of failed lines <NULL>), number edges to remove 2452, phase 0 for topology gronn_2023_v3_004_15  with tolerance 0 at 2024-10-26 08:06:37.760165+02 used_time: 139.72722
NOTICE:  Done with 1097 number of add_border_lines for _box_id 16 to temp topollogy (number of failed lines <NULL>), number edges to remove 1943, phase 0 for topology gronn_2023_v2_004_16  with tolerance 0 at 2024-10-25 10:32:06.775141+02 used_time: 81.84047
NOTICE:  Done with 1097 number of add_border_lines for _box_id 16 to temp topollogy (number of failed lines <NULL>), number edges to remove 1943, phase 0 for topology gronn_2023_v3_004_16  with tolerance 0 at 2024-10-26 08:05:30.253628+02 used_time: 77.731415
NOTICE:  Done with 2382 number of add_border_lines for _box_id 62 to temp topollogy (number of failed lines <NULL>), number edges to remove 1606, phase 0 for topology gronn_2023_v2_004_62  with tolerance 0 at 2024-10-25 10:44:16.795156+02 used_time: 120.889656
NOTICE:  Done with 2382 number of add_border_lines for _box_id 62 to temp topollogy (number of failed lines <NULL>), number edges to remove 1606, phase 0 for topology gronn_2023_v3_004_62  with tolerance 0 at 2024-10-26 08:17:28.75431+02 used_time: 117.7431
NOTICE:  Done with 1563 number of add_border_lines for _box_id 101 to temp topollogy (number of failed lines <NULL>), number edges to remove 2323, phase 0 for topology gronn_2023_v2_004_101  with tolerance 0 at 2024-10-25 10:33:07.759342+02 used_time: 100.5909
NOTICE:  Done with 1563 number of add_border_lines for _box_id 101 to temp topollogy (number of failed lines <NULL>), number edges to remove 2323, phase 0 for topology gronn_2023_v3_004_101  with tolerance 0 at 2024-10-26 08:06:40.501792+02 used_time: 110.73046
NOTICE:  Done with 993 number of add_border_lines for _box_id 102 to temp topollogy (number of failed lines <NULL>), number edges to remove 1867, phase 0 for topology gronn_2023_v2_004_102  with tolerance 0 at 2024-10-25 10:33:04.639282+02 used_time: 75.12351
NOTICE:  Done with 993 number of add_border_lines for _box_id 102 to temp topollogy (number of failed lines <NULL>), number edges to remove 1867, phase 0 for topology gronn_2023_v3_004_102  with tolerance 0 at 2024-10-26 08:06:38.228367+02 used_time: 83.239525


```

I also added the call I used for the test so we retest with the same setup.

```
CALL topo_rog_static.rog_overlay(
ARRAY[
'org_gronnstruktur_fkb.bygg_arsvers_23_v2 bygg',
'org_gronnstruktur_fkb.gsk_2023_segmenter_renska segmenter'
],
'gronn_2023_v3.gronnstruktur_23_10_2024',
4258, --srid
0.0, -- no snapto
5.0, -- min m2 area to keep
200, -- split all lines at given vertex num
true, -- _break_up_big_polygons
NULL,
ARRAY[
('0103000020A21000000100000005000000FC60C227D47F09409A564F720FB64D40FC60C227D47F0940B6D8185120874E40EBF20689D0021D40B6D8185120874E40EBF20689D0021D409A564F720FB64D40FC60C227D47F09409A564F720FB64D40','004')::rog_input_boundary
 ] , -- run IN 22 area blocks
   true, -- do_qualitycheck_on_final_reseult
   --  Method 1 topology.TopoGeo_addLinestring
   --  Method 2 topology.TopoGeo_LoadGeometry
   --  Method 3 select from tmp toplogy into master topology
   3,
55, -- _max_parallel_jobs int, -- this is the max number of paralell jobs to run. There must be at least the same number of free cinnections
2000, -- _max_rows_in_each_cell int, -- this is the max number rows that intersects with box before it's split into 4 new boxes
null,
ARRAY[
  NULL, -- no cleanup for org_gronnstruktur_fkb.bygg_arsvers_23
  rog_clean_edge_type_func(
    0.00003,
    NULL,
    ('clname','{treeLayer,fieldLayer,trees,bushLayer,FKB_bygg,greyArea}', '{0.00001,0.00001,0.00001,0.00001,0.00001,0.000025}')::rog_clean_simplify_attribute_values
  )
]
);

```

**Comment by larsop** [comment]
Here is the result from a bigger run which involved 40 mill edges

```
select max(edge_id) from gronn_2023_v5_004.edge_data;
   max    
----------
 40697226
(1 row)

```

The final topology has this stats after it's cleanup 
```

SELECT topology.topologysummary('gronn_2023_v5_004');
                            topologysummary                             
------------------------------------------------------------------------
 Topology gronn_2023_v5_004 (id 208067, SRID 4258, precision 0)        +
 24385151 nodes, 30737430 edges, 7040920 faces, 0 topogeoms in 0 layers+
 Layer 1, type Lineal (2), 0 topogeoms                                 +
  Deploy: gronn_2023_v5_004.edge_attributes.geom                       +
 Layer 2, type Polygonal (3), 0 topogeoms                              +
  Deploy: gronn_2023_v5_004.face_attributes.geom                       +
 

```

The runtime was down from `Time: 70258520.205 ms (19:30:58.520)` to `Time: 68425229.686 ms (19:00:25.230)` using this code.

**Comment by strk** [comment]
Ok merged in master branch as 020dc9366e28a62bda661eb6db6a95ac59f91294
Thank you @bjornharrtell and @larsop for the tests
